mesonpy/_editable.py

```diff
@@ -0,0 +1,155 @@
+        with self.open('rb') as strm:
+            return strm.read()  # type: ignore[return-value]
+
+    def read_text(self, encoding: Optional[str] = None) -> str:
+        """
+        Read contents of self as text
+        """
+        with self.open(encoding=encoding) as strm:
+            return strm.read()
+
+    def __truediv__(self, child: StrPath) -> resources_abc.Traversable:
+        """
+        Return Traversable child in self
+        """
+        return self.joinpath(child)
+
+    # backport from 3.12
+
+    def joinpath(self, *descendants: StrPath) -> resources_abc.Traversable:
+        if not descendants:
+            return self
+        names = itertools.chain.from_iterable(
+            path.parts for path in map(pathlib.PurePosixPath, descendants)
+        )
+        target = next(names)
+        matches = (
+            traversable
+            for traversable in self.iterdir()
+            if traversable.name == target
+        )
+        try:
+            match = next(matches)
+        except StopIteration:
+            raise ImportError(
+                'Target not found during traversal.', target, list(names)
+            )
+        return match.joinpath(*names)
+
+
+class EditableReader(resources_abc.TraversableResources):  # type: ignore[misc]
+    def __init__(self, path: EditablePath) -> None:
+        self._path = path
+
+    def files(self) -> resources_abc.Traversable:
+        return self._path
+
+
+class EditableLoader(importlib.machinery.SourceFileLoader):
+    def __init__(self, fullname: str, path: EditablePath) -> None:
+        super().__init__(fullname, fullname)
+        self._fullname = fullname
+        self._path = path
+
+    def __repr__(self) -> str:
+        return f'{self.__class__}({self._path})'
+
+    def get_data(self, path: Union[str, bytes]) -> bytes:
+        if path != self._fullname:
+            raise ImportError(f'Requesting data from unexpected path `{path!r}` in {self}')
+        return self._path.read_bytes()
+
+    def get_resource_reader(self) -> importlib.abc.ResourceReader:
+        return EditableReader(self._path)
+
+
+class MesonpyFinder(importlib.abc.MetaPathFinder):
+    def __init__(self, editable: Editable) -> None:
+        self._editable = editable
+        self._base_path = EditablePath(self._editable)
+
+    def __repr__(self) -> str:
+        return f'{self.__class__}({self._editable})'
+
+    def find_spec(
+        self,
+        fullname: str,
+        path: Optional[Sequence[Union[str, bytes]]],
+        target: Optional[ModuleType] = None,
+    ) -> Optional[importlib.machinery.ModuleSpec]:
+        if not fullname.startswith('aaaa'):
+            return None
+        spec_path = self._base_path.joinpath(*fullname.split('.'))
+        return importlib.util.spec_from_loader(fullname, EditableLoader(fullname, spec_path))
+
+
+def install(path: str) -> None:
+    editable = Editable(os.path.abspath(path))
+    sys.meta_path.append(MesonpyFinder(editable))
```

Comment:
Would a prepend be more appropriate here?

Reply:
Doesn't matter much, but my reasoning for simply appending was to not slow down any of the other imports. But this is now outdated given our new approach.